funding/batch.go

```diff
@@ -0,0 +1,528 @@
+	// local database, graph and arbitrator state.
+	ChannelAbandoner ChannelAbandoner
+
+	// WalletKitServer is an instance of the wallet kit sub server that can
+	// handle PSBT funding and finalization.
+	WalletKitServer WalletKitServer
+
+	// Wallet is an instance of the internal lightning wallet.
+	Wallet Wallet
+
+	// NetParams contains the current bitcoin network parameters.
+	NetParams *chaincfg.Params
+
+	// Quit is the channel that is selected on to recognize if the main
+	// server is shutting down.
+	Quit chan struct{}
+}
+
+// Batcher is a type that can be used to perform an atomic funding of multiple
+// channels within a single on-chain transaction.
+type Batcher struct {
+	cfg *BatchConfig
+
+	channels    []*batchChannel
+	lockedUTXOs []*walletrpc.UtxoLease
+
+	didPublish bool
+}
+
+// NewBatcher returns a new batch channel funding helper.
+func NewBatcher(cfg *BatchConfig) *Batcher {
+	return &Batcher{
+		cfg: cfg,
+	}
+}
+
+// BatchFund starts the atomic batch channel funding process.
+//
+// NOTE: This method should only be called once per instance.
+func (b *Batcher) BatchFund(ctx context.Context,
+	req *lnrpc.BatchOpenChannelRequest) ([]*lnrpc.PendingUpdate, error) {
+
+	label, err := labels.ValidateAPI(req.Label)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse and validate each individual channel.
+	b.channels = make([]*batchChannel, 0, len(req.Channels))
+	for idx, rpcChannel := range req.Channels {
+		// If the user specifies a channel ID, it must be exactly 32
+		// bytes long.
+		if len(rpcChannel.PendingChanId) > 0 &&
+			len(rpcChannel.PendingChanId) != 32 {
+
+			return nil, fmt.Errorf("invalid temp chan ID %x",
+				rpcChannel.PendingChanId)
+		}
+
+		var pendingChanID [32]byte
+		if len(rpcChannel.PendingChanId) == 32 {
+			copy(pendingChanID[:], rpcChannel.PendingChanId)
+
+			// Don't allow the user to be clever by just setting an
+			// all zero channel ID, we need a "real" value here.
+			if pendingChanID == emptyChannelID {
+				return nil, fmt.Errorf("invalid empty temp " +
+					"chan ID")
+			}
+		} else if _, err := rand.Read(pendingChanID[:]); err != nil {
+			return nil, fmt.Errorf("error making temp chan ID: %v",
+				err)
+		}
+
+		fundingReq, err := b.cfg.RequestParser(&lnrpc.OpenChannelRequest{
+			SatPerVbyte:        uint64(req.SatPerVbyte),
+			NodePubkey:         rpcChannel.NodePubkey,
+			LocalFundingAmount: rpcChannel.LocalFundingAmount,
+			PushSat:            rpcChannel.PushSat,
+			TargetConf:         req.TargetConf,
+			Private:            rpcChannel.Private,
+			MinHtlcMsat:        rpcChannel.MinHtlcMsat,
+			RemoteCsvDelay:     rpcChannel.RemoteCsvDelay,
+			MinConfs:           req.MinConfs,
+			SpendUnconfirmed:   req.SpendUnconfirmed,
+			CloseAddress:       rpcChannel.CloseAddress,
+			FundingShim: &lnrpc.FundingShim{
+				Shim: &lnrpc.FundingShim_PsbtShim{
+					PsbtShim: &lnrpc.PsbtShim{
+						PendingChanId: pendingChanID[:],
+						NoPublish:     true,
+					},
+				},
+			},
```

Comment:
We would then also thread through the `commitment_type` field here as well: https://github.com/lightningnetwork/lnd/blob/d263a01a2d3bb3570538418a326a8ed165e500a8/rpcserver.go#L1916 (example of where it's read to eventually give to the funding manager).